src/renderers/WebGLRenderer.js

```diff
@@ -861,6 +861,162 @@ function WebGLRenderer( parameters = {} ) {
 
 	};
 
+	// compileAsync
+
+	this.compileAsync = function ( scene, targetScene = null ) {
+
+		// If no explicit targetScene was given use the scene instead
+		if ( ! targetScene ) {
+
+			targetScene = scene;
+
+		}
+
+		currentRenderState = renderStates.get( targetScene );
+		currentRenderState.init();
+
+		renderStateStack.push( currentRenderState );
+
+		let foundScene = scene === targetScene;
+
+		// Gather lights from both the scene and the new object that will be added
+		// to the scene.
+		targetScene.traverseVisible( function ( object ) {
+
+			if ( object === scene ) {
+
+				foundScene = true;
+
+			}
+
+			if ( object.isLight ) {
+
+				currentRenderState.pushLight( object );
+
+				if ( object.castShadow ) {
+
+					currentRenderState.pushShadow( object );
+
+				}
+
+			}
+
+		} );
+
+		// If the scene wasn't already part of the targetScene, add any lights it
+		// contains as well.
+		if ( ! foundScene ) {
+
+			scene.traverseVisible( function ( object ) {
+
+				if ( object.isLight ) {
+
+					currentRenderState.pushLight( object );
+
+					if ( object.castShadow ) {
+
+						currentRenderState.pushShadow( object );
+
+					}
+
+				}
+
+			} );
+
+		}
+
+		currentRenderState.setupLights( _this.physicallyCorrectLights );
+
+		const compiling = new Set();
+
+		// Only initialize materials in the new scene, not the targetScene.
+
+		scene.traverse( function ( object ) {
```

Comment:
Could we add scene `overrideMaterial` support? It seemed to work when I did the naive thing:

```
const _compileMaterial = function ( material, object ) {
	if ( Array.isArray( material ) ) {

		for ( let i = 0; i < material.length; i ++ ) {

			const material2 = material[ i ];

			getProgram( material2, targetScene, object );
			compiling.add( material2 );

		}

	} else {

		getProgram( material, targetScene, object );
		compiling.add( material );

	}
};

if (scene.overrideMaterial) {
	scene.traverse( function ( object ) {
		const material = object.material;

		if ( material ) {
			_compileMaterial(scene.overrideMaterial, object);
		}
	});
} else {
	scene.traverse( function ( object ) {

		const material = object.material;

		if ( material ) {

			_compileMaterial(material, object);

		}

	} );
}
```